lib/utils/middlewares/azure_api_auth.ts

```diff
@@ -97,45 +102,68 @@ export function AzureApiAuthMiddleware(
   IAzureApiAuthorization
 > {
   return request =>
     new Promise(resolve => {
+      // get Azure userId and subscriptionId from the headers
+      // these headers get added by the Azure API Manager gateway
+      const maybeUserId = toNonEmptyString(request.header("x-user-id"));
+      const maybeSubscriptionId = toNonEmptyString(
+        request.header("x-subscription-id")
+      );
+
+      if (maybeUserId.isEmpty || maybeSubscriptionId.isEmpty) {
+        // we cannot proceed unless we cannot associate the request to a
+        // valid user and a subscription
+        return resolve(left(ResponseErrorForbiddenAnonymousUser));
+      }
+
+      const userId = maybeUserId.get;
+      const subscriptionId = maybeSubscriptionId.get;
+
       // to correctly process the request, we must associate the correct
       // authorizations to the user that made the request; to do so, we
       // need to extract the groups associated to the authenticated user
       // from the x-user-groups header, generated by the Azure API Management
       // proxy.
-      const groupsHeader = request.header("x-user-groups");
-      option(groupsHeader)
+      const maybeGroupsHeader = toNonEmptyString(
+        request.header("x-user-groups")
+      );
+      const maybeGroups = maybeGroupsHeader
         .map(getGroupsFromHeader) // extract the groups from the header
-        .filter(hs => hs.size > 0) // filter only if set of groups is non empty
-        .map(groups => {
-          // now we have some valid groups that the users is in
-
-          // helper that checks whether the user is part of a specific group
-          const userHasOneGroup = (name: UserGroup) => groups.has(name);
-          // whether the user is part of at least an allowed group
-          const userHasAnyAllowedGroup =
-            Array.from(allowedGroups).findIndex(userHasOneGroup) > -1;
-
-          if (userHasAnyAllowedGroup) {
-            // the user is allowed here
-            const authInfo: IAzureApiAuthorization = {
-              groups,
-              kind: "IAzureApiAuthorization"
-            };
-
-            resolve(right(authInfo));
-          } else {
-            // or else no valid groups, the user is not allowed here
-            resolve(left(ResponseErrorForbiddenNotAuthorized));
-          }
-        })
-        .getOrElse(() => {
-          // or else no valid groups
-          resolve(left(ResponseErrorForbiddenNoAuthorizationGroups));
-        });
+        .filter(hs => hs.size > 0); // filter only if set of groups is non empty
+
+      if (maybeGroups.isEmpty) {
+        // the use as no valid authorization groups
+        return resolve(left(ResponseErrorForbiddenNoAuthorizationGroups));
+      }
+
+      // now we have some valid groups that the users is part of
+      const groups = maybeGroups.get;
+
+      // helper that checks whether the user is part of a specific group
+      const userHasOneGroup = (name: UserGroup) => groups.has(name);
```

Comment:
maybe it's better to put this at the module level and / or make a dedicated user module with this kind of utilities.

Reply:
not sure what you mean, how would you make this helper generic?
it's just a call the `Set#has` method on the locally scoped `groups`